### DISCUSSION TITLE
WikiJS on Docker with Digital Ocean Database Cluster

### DISCUSSION THREAD

**Discussion by DevilYanYan** [requarks/wiki#6253] [Help / Questions] [answered]
I'm trying to run a Dockerized WikiJS that connects to an external PostgreSQL Server.

First I attempted to run a container connected to a simple external PostgreSQL Server, and was successful running:

```
docker run --name wiki \
    --restart unless-stopped \
    -e DB_TYPE=postgres \
    -e DB_HOST=ip.to.my.psql \
    -e DB_PORT=5432 \
    -e DB_USER=wiki \
    -e DB_PASS=password_here \
    -e DB_NAME=wiki \
    --network tools \
    -d ghcr.io/requarks/wiki:2
```

Then for a Production Deploy, now that I had it working for the previous command, was to do the same Container but to connect to my Digital Ocean PostgreSQL Cluster, there's 2 main differences, it requires SSL so I tried first with the Environment Variable DB_SLL=true and got a missing CA File Error, so I downloaded the CA File from my DO Cluster, and added the Environment Variable DB_SSL_CA=no-preffix-no-suffix-all-one-line:

```
docker run --name wiki-digital-ocean \
    --restart unless-stopped \
    -e DB_TYPE=postgres \
    -e DB_HOST=postgresql.db.ondigitalocean.com \
    -e DB_PORT=25060 \
    -e DB_USER=wiki \
    -e DB_PASS=password_here \
    -e DB_NAME=wiki \
    -e DB_SSL=true \
    -e DB_SSL_CA=no-preffix-no-suffix-all-one-line \
    --network tools \
    -d ghcr.io/requarks/wiki:2
```

Now I get on the WikiJS Container Log the following error:

```
error: Database Connection Error: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
```

Thinking it was probably a matter of Network Block, I decided to spin up a PostgreSQL Container on the same Docker Host and same Docker Network, and try to connect to it and it connected no issues:

```
docker exec -it postgres psql -U wiki -h postgresql.db.ondigitalocean.com -p 25060 -d wiki --set=sslmode=require
Password for user wiki:
psql (15.2 (Debian 15.2-1.pgdg110+1), server 14.5)
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, compression: off)
Type "help" for help.

wiki=>
```

What am I missing? I double checked and the Environmental Variable for the CA is correct, I removed the Start Preffix and the End Suffix as well as I made sure it was one line, if I inspect it on Portainer or Docker it looks correct.

I would really like to deploy it with the DO Cluster.

**Comment by NGPixel** [comment]
Have you allowed your droplet IP in your db cluster? You should be using the internal private hostname for your db cluster, not the external one.

**Reply by DevilYanYan** [reply]
Thanks for the suggestions and taking the time to help me,

Yes to both, I'm using the private host name for the Database Cluster and yes I have allowed the Droplet in the Trusted Sources.

**Comment by DevilYanYan** [comment]
As additional info, I have a different Droplet running a Stack of Docker Containers with NGinX (for reverse Proxy) and Odoo, they are on their own Network, the NGinX Container have ports 80 and 443 exposed, the Odoo Container doesn't have it's ports exposed and is able to connect to the same Database Cluster without any issues.

Both Droplets, the one where WikiJS is and the one with the NGinX and Odoo Stack are on the same VPC Network as well as the Database Cluster, and have same global settings and packages, the only differences are:

- The Docker Containers each one is running
- The Specifications, the WikiJS is a 2 CPU 4 GB of RAM Droplet, the Containers Stack one is 4 CPU 8 GB
- No Firewall is up while testing connectivity on both instances, Server Internal Firewall (UFW) or Digital Ocean's Firewall

**Comment by DevilYanYan** [comment, marked as answer]
Found the issue,

Do not try to pass the DB_SSL_CA as a one liner from Docker Run or Docker Compose directly on a variable like:

```
-e DB_SSL_CA=VeryLongLine.
```

Instead create an Environment Variables file, example:

`wikijs.env`:

```
DB_TYPE=postgres
DB_HOST=private-postgresql.db.ondigitalocean.com
DB_PORT=25060
DB_USER=wiki
DB_PASS=password_here
DB_NAME=wiki
DB_SSL=true
DB_SSL_CA=VeryLongLine
```

Very important, DO NOT use any " ' " to open or close any of the values, it will fail.

Then just run:

```
docker run --name wiki-digital-ocean \
    --restart unless-stopped \
    --env-file wikijs.env \
    -d ghcr.io/requarks/wiki:2
```

Also worked once you turn this into a docker-compose.yml
